src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProject.cs

```diff
@@ -1639,15 +1662,66 @@ public bool ContainsFile(string fullPath)
                 }
             }
 
-            public void ProcessFileChange(string filePath)
-                => ProcessFileChange(filePath, filePath);
+            public async ValueTask ProcessRegularFileChangesAsync(ImmutableArray<string> filePaths)
+            {
+                using (await _project._gate.DisposableWaitAsync().ConfigureAwait(false))
+                {
+                    // If our project has already been removed, this is a stale notification, and we can disregard.
+                    if (_project.HasBeenRemoved)
+                    {
+                        return;
+                    }
+
+                    var documentsToChange = new List<(DocumentId, TextLoader)>(capacity: filePaths.Length);
+
+                    foreach (var filePath in filePaths)
+                    {
+                        if (_documentPathsToDocumentIds.TryGetValue(filePath, out var documentId))
+                        {
+                            // We create file watching prior to pushing the file to the workspace in batching, so it's
+                            // possible we might see a file change notification early. In this case, toss it out. Since
+                            // all adds/removals of documents for this project happen under our lock, it's safe to do this
+                            // check without taking the main workspace lock. We don't have to check for documents removed in
+                            // the batch, since those have already been removed out of _documentPathsToDocumentIds.
+                            if (!_documentsAddedInBatch.Any(d => d.Id == documentId))
```

Comment:
```suggestion
                            if (!_documentsAddedInBatch.Any(static (d, documentId) => d.Id == documentId, documentId))
```

Reply:
This is on an ImmutableArray.Builder, so we don't have that helper here, unless there's some other overload somewhere?